enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#![allow(dead_code)]
+
+use crate::image::ImageDetails;
+use crate::storage::OciStorage;
+use crate::{EnclaveBuildError, Result};
+
+use tokio::runtime::Runtime;
+
+pub struct OciImageManager {
+    /// Name of the container image.
+    image_name: String,
+    /// Have the storage as an option for the CLI to continue running if there is a storage creation
+    /// error (the image will simply be pulled in this case, instead of being fetched from storage)
+    storage: Option<OciStorage>,
+}
+
+impl OciImageManager {
+    /// When calling this constructor, it also tries to initialize the storage at the default path.
+    /// If this fails, the ImageManager is still created, but the 'storage' field is set to 'None'.
+    pub fn new(image_name: &str) -> Self {
+        // Add the default ":latest" tag if the image tag is missing
+        let image_name = check_tag(image_name);
+
+        // The docker daemon is not used, so a local storage needs to be created
+        // Get the default storage root path
+        let root_path = match OciStorage::get_default_root_path() {
+            Ok(path) => path,
+            Err(_) => {
+                // If the storage root path could not be determined, then the storage can not be initialized
+                return Self {
+                    image_name,
+                    storage: None,
+                };
+            }
+        };
+
+        // Try to create/read the storage
+        let storage = match OciStorage::new(&root_path) {
+            Ok(manager) => Some(manager),
+            Err(err) => {
+                // If the storage could not be created, log the error
+                eprintln!("{:?}", err);
+                None
+            }
+        };
+
+        Self {
+            image_name,
+            storage,
+        }
+    }
+
+    /// Returns a struct containing image metadata.
+    ///
+    /// If the image is stored correctly, the function tries to fetch the image from the storage.
+    ///
+    /// If the image is not stored or a storage was not created (the 'storage' field is None),
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn get_image_details(&mut self, image_name: &str) -> Result<ImageDetails> {
```

Comment:
Why do we need a mutable reference? From the function name, I would imagine self is not mutated. I think the problem comes from the fact that we are doing more things than we should in a single function, maybe we can refactor a bit (do the storing separately). Is that doable?

Reply:
The idea of this image manager is to hide the storage implementation while allowing the EIF to be built even if we don't have the image stored. We have to mutate the storage at some point and I think it complicates things if we move it higher in the workflow.